tools/bigquery-zos-mainframe-connector/gszutil/src/main/scala/com/google/cloud/bqsh/cmd/Extract.scala

```diff
@@ -0,0 +1,151 @@
+          allowLargeResults = cfg.allowLargeResults, destinationTable = cfg.destinationTable, dryRun = cfg.dryRun,
+          maximumBytesBilled = cfg.maximumBytesBilled, requireCache = cfg.requireCache, useCache = cfg.useCache,
+          useLegacySql = cfg.useLegacySql, projectId = cfg.projectId, datasetId = cfg.datasetId,
+          location = cfg.location, debugMode = cfg.debugMode, batch = cfg.batch, sync = true,
+          jobProperties = cfg.jobProperties, replace = cfg.replace)
+        val (jobId, job) = BqQueryJobExecutor(bq, queryConfig, zos).execute(query, Query.configureQueryJob)
+        val table = job.getConfiguration[QueryJobConfiguration].getDestinationTable
+        val rowCount = BQ.rowCount(bq, table)
+        val stats = job.getStatistics[QueryStatistics]
+        val result = extract(table, bq, cfg, zos)
+
+        // Publish results
+        if (cfg.statsTable.nonEmpty) {
+          val statsTable = BQ.resolveTableSpec(cfg.statsTable, cfg.projectId, cfg.datasetId)
+          val jobId = BQ.genJobId(cfg.projectId, cfg.location, zos, "extract")
+          val tblspec = s"${statsTable.getProject}:${statsTable.getDataset}.${statsTable.getTable}"
+          logger.debug(s"Writing stats to $tblspec, with jobId ${jobId}")
+          StatsUtil.retryableInsertJobStats(zos, jobId, bq, statsTable, jobType = "export", recordsOut = result.activityCount)
+        }
+
+        logger.info(s"Extract completed, extracted=${rowCount} rows, " +
+          s"time took: ${(System.currentTimeMillis() - startTime) / 1000} seconds.")
+        result
+      } else {
+        // Extract contents of a table
+        val table: TableId = BQ.resolveTableSpec(cfg.sourceTable, cfg.projectId, cfg.datasetId)
+        logger.info(s"Extract from source table: $table")
+        val rowCount = BQ.rowCount(bq, table)
+        val result = extract(table, bq, cfg, zos)
+
+        // Publish results
+        if (cfg.statsTable.nonEmpty) {
+          val statsTable = BQ.resolveTableSpec(cfg.statsTable, cfg.projectId, cfg.datasetId)
+          val jobId = BQ.genJobId(cfg.projectId, cfg.location, zos, "extract")
+          val tblspec = s"${statsTable.getProject}:${statsTable.getDataset}.${statsTable.getTable}"
+          logger.debug(s"Writing stats to $tblspec, with jobId ${jobId}")
+          StatsUtil.retryableInsertJobStats(zos, jobId, bq, statsTable, jobType = "export", recordsOut = result.activityCount)
+        }
+
+        logger.info(s"Extract completed, extracted=${rowCount} rows, " +
+          s"time took: ${(System.currentTimeMillis() - startTime) / 1000} seconds.")
+        result
+      }
+
+    } catch {
+      case e: Throwable =>
+        val msg = "Extract failed: " + e.getMessage + "\n"
+        logger.error(msg, e)
+        Result.Failure(msg)
+    }
+  }
+
+  def extract(table: TableId, bq: BigQuery, cfg: ExtractConfig, zos: MVS): Result = {
+    val jobId = BQ.genJobId(cfg.projectId, cfg.location, zos, "extract", Util.generateHashString)
```

Comment:
why are we generating a custom BQ job id?

Reply:
It's not strictly necessary because BigQuery will autogenerate an ID, but specifying our own makes it a human readable one that shows the z/OS job and step name.